p11-kit/proxy.c

```diff
@@ -279,29 +277,59 @@ proxy_list_slots (Proxy *py, Mapping *mappings, unsigned int n_mappings)
 
 		if (count > 0) {
 			Mapping *new_mappings;
+			bool *is_new_mapping;
+			CK_SLOT_ID max_id = 0;
 
+			is_new_mapping = calloc (count, sizeof(bool));
+			return_val_if_fail (is_new_mapping != NULL, CKR_HOST_MEMORY);
 			new_mappings = realloc (py->mappings, sizeof (Mapping) * (py->n_mappings + count));
 			return_val_if_fail (new_mappings != NULL, CKR_HOST_MEMORY);
 			py->mappings = new_mappings;
 
-			/* And now add a mapping for each of those slots */
+			/* Reuse the existing mapping if any */
 			for (i = 0; i < count; ++i) {
-				/* Reuse the existing mapping if any */
 				for (j = 0; j < n_mappings; ++j) {
 					/* cppcheck-suppress nullPointer symbolName=mappings */
 					/* false-positive: https://trac.cppcheck.net/ticket/9573 */
 					if (mappings[j].funcs == funcs &&
-					    mappings[j].real_slot == slots[i])
+					    mappings[j].real_slot == slots[i]) {
+						py->mappings[py->n_mappings].funcs = funcs;
+						py->mappings[py->n_mappings].real_slot = slots[i];
+						py->mappings[py->n_mappings].wrap_slot =
+							mappings[j].wrap_slot;
+						++py->n_mappings;
 						break;
+					}
 				}
-				py->mappings[py->n_mappings].funcs = funcs;
-				py->mappings[py->n_mappings].wrap_slot =
-					(n_mappings == 0 || j == n_mappings) ?
-					py->n_mappings + MAPPING_OFFSET :
-					mappings[j].wrap_slot;
-				py->mappings[py->n_mappings].real_slot = slots[i];
-				++py->n_mappings;
+				if (n_mappings == 0 || j == n_mappings)
+					is_new_mapping[i] = true;
 			}
+
+			/* Remember the largest ID assigned */
+			for (j = 0; j < n_mappings; ++j) {
+				if (mappings[j].wrap_slot > max_id)
+					max_id = mappings[j].wrap_slot;
+			}
+
+			/* The ID for the next new mapping will be greater than any
+			 * previously assigned ID */
+			++max_id;
+			if (max_id > MAPPING_OFFSET)
```

Comment:
any reason to subtract MAPPING_OFFSET while it is added back below?

Reply:
The reason is to keep the ID's sequential and not waste ``MAPPING_OFFSET`` ID's. But I agree it wasn't implemented in the best way. I think the best would be to keep the last ID assigned in the proxy structure.